\documentclass[12pt,letterpaper]{article}
\usepackage[a4paper, total={7in, 10in}]{geometry}
\renewcommand{\familydefault}{\sfdefault}
\usepackage{graphicx}
\usepackage{helvet}
\usepackage{authblk}
\usepackage{hyperref}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{orcidlink}
\usepackage{booktabs}
\usepackage{tabularx}
\usepackage{makecell}
\usepackage[super,comma,sort&compress]
   {natbib}\bibliographystyle{numbered}
\usepackage[right]{lineno} \linenumbers

\graphicspath{ {../figures/} }

\makeatletter
\renewcommand{\maketitle}{\bgroup\setlength{\parindent}{0pt}
\begin{flushleft}
  \textbf{\@title}

  \@author
\end{flushleft}\egroup}
\makeatother

\title{Document S1}
\date{}

\begin{document}

\maketitle

\section*{Supplementary Text}

\subsection*{Identification of robust recombination events}
Before characterizing recombination events in detail, we sought to remove probable artifacts.
Systematic errors in SARS-CoV-2 sequences associated with sequencing protocols have been documented (Turakhia et al., 2022),
which can confound recombination detection (De Maio et al., 2024).
By inspecting copying patterns at sites that differ between recombinant parents,
we find a convincing signal when treating adjacent and near-adjacent sites as a single locus.
Artifactual recombination events can be characterized as
having fewer than four such loci supporting the left or right parent on either side of a suggested breakpoint (STAR Methods).
Of the 929 recombination events in the ARG, this measure identifies 543 as artifactual,
leaving 386 “robust” events (Figure S1A).
The copying patterns depicted in Figure S1B illustrate the genomic evidence for
three recombination events identified as artifactual, compared to a marginally robust one.

Evidence that our simple filter effectively flags artifacts can be seen in the clear association between
the AmpliSeq V1 kit and recombination events that we identify as artifactual.
This primer scheme was used to sequence XX of the samples associated with an artifactual recombination event,
in contrast to 15 of the “robust” events (Figure 2A), strongly implicating it as a common source of problematic data.
Through further inspection, we observed that many artifactual recombination events
contain stretches of missing data characteristic of amplicon dropouts (REF).

\subsection*{Analysis of Pango lineage ``origination'' events}
Notebook

To analyse how well the ARG reflects the phylogenetic structure implicit in the Pango lineage naming system,
we considered the Pango assignments generated by Pangolin on the alignments for each node in the ARG (STAR Methods).
If the ARG perfectly reflected the Pango lineage structure,
each lineage would form a clade descending from a single originating node with that label.
We identified putative origination nodes among those labeled with a given lineage as the node
with the maximum number of descendant samples and the earliest time (if there is a tie).

Of the 2058 distinct Pango lineages in the ARG,
1469 of these (comprising 743922 samples) match perfectly,
with unique origination events in the ARG
where all samples assigned a given lineage descend from the first node assigned to that lineage.
A further 310 lineages (398391 samples) match perfectly
when we count the descendants of the parent of the first node
(accounting for polytomies in which multiple originating nodes for a given lineage are siblings).
We then have 255 lineages (996490 samples)
where the difference in the number of descendants of the first node's parent is < 100.
The remaining 25 lineages (344079) are dominated by a few large lineages such as BA.1.1 (155595 samples) and AY.4.2 (54607 samples)
which have multiple non-sibling origins within the ARG.

\subsection*{Additional analyses of the Pango Xs}
\subsubsection*{Multiple origins}
\subsubsection*{XS}
https://github.com/jeromekelleher/sc2ts-paper/issues/287
In the ARG, the XS samples appear to have originated from two sequential recombination events.
However, the second event is likely an artifact,
as the XS sample associated with this event failed the minimum net number of supporting loci.
In addition, this sample was sequenced using the AmpliSeq V1 kit,
which was also used to sequence eight other samples descending from the same event.
We excluded this likely artifactual event from further analysis and
consider XS to have emerged from a single recombinant origin.

\subsubsection*{XM}
https://github.com/jeromekelleher/sc2ts-paper/issues/284
Possibly related to deletions.

\subsubsection*{Discordant parents}
\subsubsection*{XBB}

\subsubsection*{XBF}

\subsubsection*{XBH}

\subsection*{Multiple breakpoints}
Seven of the Pango Xs present in the Viridian v04 dataset have more than one breakpoint,
with the supporting evidence documented in the associated Pango designation issues.
Five of these seven Pango Xs were detected as recombinants by sc2ts.
Three of them (XAY, XBC, XBT) had more than one breakpoint,
but were not added to the ARG due to high HMM cost and
a lack of sample support to be incorporated as retrospective sample groups.
However, the other two (XAC and XAK) had only one breakpoint.

\subsubsection*{XAY}
https://github.com/jeromekelleher/sc2ts-paper/issues/490

\subsubsection*{XBC}
https://github.com/jeromekelleher/sc2ts-paper/issues/486
Sc2ts detected XBC as a three-breakpoint recombinant,
correctly identifying the parental Pango lineages.
The inferred breakpoint intervals mostly overlap with those identified in its designation issue.
XBC was not added to the ARG due to high HMM costs and inadequate sample support
(within a seven-day retrospective time window).

\subsubsection*{XBT}
https://github.com/jeromekelleher/sc2ts-paper/issues/487
Sc2ts detected XBT as a two-breakpoint recombinant,
correctly identifying the parental Pango lineages.
The inferred breakpoint intervals are identical to those documented in its designation issue.
XBT was not added to the ARG due to high HMM cost and low sample support
(only one QC-passing sample in the Viridian v04 dataset).

\subsubsection*{XAC}
https://github.com/jeromekelleher/sc2ts-paper/issues/357
Evidence of the second breakpoint reported in its designation issue relies on
the presence of a large deletion inherited from BA.2*, which is currently treated as missing data by sc2ts.
The XAC samples share a recombination node with the samples of four other Pango Xs
which do not have this deletion (Figure 2C).

\subsubsection*{XAK}
https://github.com/jeromekelleher/sc2ts-paper/issues/484
Our results agree with its designation issue in terms of the parental Pango lineages and breakpoint intervals.
XAK was not added to the ARG due to HMM costs (9 and 10) and insufficient sample support (only two samples).
Interestingly, our results suggest that XAK might be a product of two sequential recombination events.

\subsubsection*{Non-recombinants}
\subsubsection*{XN and XAU}
https://github.com/jeromekelleher/sc2ts-paper/issues/358
https://github.com/jeromekelleher/sc2ts-paper/issues/348
Both of these putative recombinants have been defined on the basis of a RHS from BA.2 and a LHS from BA.1.
In the case of XN, the supposed BA.1 mutations are [C241T, A2832G, C3037T].
In the case of XAU, this set also includes C2470T, so [C241T, C2470T, A2832G, C3037T].
However, both BA.1 and BA.2 are now defined as also having C241T and C3037T.
In addition, a very large number of BA.2 samples also have A2832G, and three BA.2-designated samples also have C2470T.
This means that there is essentially no longer any evidence for a BA.1 parent for XN, and
only one uncertain mutation that suggests a BA.1 parent for XAU.

\subsubsection*{XAS}
https://github.com/jeromekelleher/sc2ts-paper/issues/340
This was proposed as a recombination between BA.5 (LHS) and BA.2 (RHS).
However, we find that this is only 2 mutations different from e.g. ERR9929846 which is a BA.4 sample.
In fact, XAS only differs from the BA.4 root by a single mutation, C27945T,
which was identified in the original designation issue as an important ORF truncation mutation also present in a few BA.2.65 samples.
It is possible that C27945T was acquired by recombination, this is clearly below our threshold for considering something to be a recombinant.

\subsubsection*{XB}
https://github.com/jeromekelleher/sc2ts-paper/issues/332
XB was proposed to have originated from recombination between B.1.634 (left parent) and B.1.631 (right parent).
None of the samples in Viridian v04 belonging to these Pango lineages were integrated into the ARG due to high HMM costs.

\subsubsection*{XP}
https://github.com/jeromekelleher/sc2ts-paper/issues/345

\subsubsection*{XAJ}
https://github.com/jeromekelleher/sc2ts-paper/issues/352

\subsubsection*{XAZ}
https://github.com/jeromekelleher/sc2ts-paper/issues/356
Proposed as a recombinant of BA.2.5 on the left and BA.5 on the right, with 3 supporting sites (C2232T, C3317T, T3358C).
In sc2ts this is very close to being a recombinant of BA.2.5 (ERR9615610) on the left,
as not only are the 3 mutations present in this sample,
but there is another site (C1912T apparently missed in the designation) which supports recombination.
XAZ has C44T which is lacking in all BA.2.5 samples,
giving only three net supporting loci for a putative recombinant.
However, position 44 is often clipped, so this is not strong evidence against a recombinant, and
this is a very finely balanced case.

\subsubsection*{XAN and XAV}
https://github.com/jeromekelleher/sc2ts-paper/issues/353
Below threshold.

\subsubsection*{XBE}
This was proposed as a recombinant between BA.5,5 and BE.4.
We find two net supporting loci for a recombination breakpoint to be around 22599.

\subsection*{Analysis of the Jackson et al. (2021) recombinants}
Jackson et al. (2021) conducted detailed analyses on some early recombinants
involving B.1.1.7 (Alpha), including the first designated Pango X lineage, XA.
The authors first performed a targeted search for candidate recombinants by
scanning sampled sequences for sequence motifs associated with Alpha, followed
by an analysis to reduce the list of plausible recombinants displaying signs of
onward transmission. Here we compared our sc2ts detection results of these
recombinants with the detection results reported by Jackson et al.

% Data derived from the notebooks/jackson_recombinants.ipynb notebook
Jackson et al.\ described four groups of recombinants (named A to D) and
four singleton recombinants.
Of the 16 samples analysed by Jackson et al, 12 were included in the ARG;
the remaining 4 were excluded because the sequences did not pass the 
QC filtering criteria, or due to a high HMM cost. Groups A-D are summarised 
in Table~\ref{tab:jackson} and show excellent concordance in terms 
of the parental Pango lineages and breakpoint intervals. All four groups
are directly associated with strongly supported recombination events 
in the ARG in terms of the number of averted mutations 
(A: 16, B: 6, C: 30, D: 18). 
The singleton CAMC-CBA018 has parents
B.1.177+B.1.1.7 and breakpoint interval 16,177--20,133 
in the ARG 
and is strongly supported by 14 averted mutations.
The Jackson et al.\ results are almost identical with the same parent 
lineages, and breakpoint interval of interval 20,390--21,256.
The singleton MILK-103C712 
has parents B.1.177.16+B.1.1.7 and a breakpoint interval of 
24915--27972 
in the ARG, again strongly supported by 11 averted mutations.
The Jackson et al.\ results for this lineage 
has parents B.1.177.17+B.1.1.7 but two breakpoints
409--446 and 26,802--27,878.


\begin{table} \centering
\caption{\label{tab:jackson}Comparison of recombination breakpoint
intervals and parent lineages for Groups A-D
reported by Jackson et al.\ with the corresponding
recombination events in the sc2ts ARG.
The second column gives the number of sequences in the group.
See the text for details of the groups and the sequences included.
limited to the samples considered by Jackson et al.
The 3SEQ coordinates
reported by Jackson et al.\ have been altered as follows: we
add one to both left and right coordinates to correspond to the
\texttt{tskit} definition of inheritance on either side of a breakpoint,
and add one to the right coordinate to make the intervals
right-exclusive. }
\begin{tabular}{lllr@{--}lr@{+}l}
\toprule
Group        & Sequences & Method & \multicolumn{2}{c}{Interval}
    & \multicolumn{2}{r}{Parent lineages} \\
\midrule
A (XA)       & 2   & Jackson        &  21,256&21,615 & B.1.177&B.1.1.7 \\
             &     &\texttt{sc2ts} &  20,411&21,765 & B.1.177.18&B.1.1.7 \\
\cmidrule{3-7}
B            & 2   & Jackson        &  6,529&6,955 & B.1.36.28&B.1.1.7  \\
             &     &\texttt{sc2ts} &   6,529&6,954 & B.1.36.28&B.1.1.7  \\
\cmidrule{3-7}
C            & 3   &Jackson         &  25,997&27,443 &  B.1.1.7&B.1.221 \\
             &     & \texttt{sc2ts} &  25,997&27,972 &  B.1.1.7&B.1.221.1 \\
\cmidrule{3-7}
D            & 3   & Jackson        &  21,576&23,064 &  B.1.36.17&B.1.1.7 \\
             &     & \texttt{sc2ts} &  22,445&23,063 &  B.1.36.39&B.1.1.7 \\
\bottomrule
\end{tabular}
\end{table}


\section*{Supplementary Figures}

\subsection*{Figure S1: Classification of “robust” and artifactual recombination events.}
(A) Scatterplot of recombination events by the net number of supporting loci for the left parent (x-axis) and for the right parent (y-axis).
The shaded regions highlight artifactual recombination events,
which have fewer than four net supporting loci on one side or both sides of the suggested breakpoint.
Colors indicate the primer scheme used for sequencing.
The pie charts show breakdowns of the recombination events by primer scheme per quadrant.
Recombination events associated with the origins of Pango Xs are labeled (classified in Figure 2).
(B) Copying patterns of an illustrative recombination event picked from each quadrant.

% JK commenting these out as we don't have the figures and want this to compile
% \noindent\includegraphics[width=0.85\linewidth]{FigureS1.png}

\subsection*{Figure S2: Comparison of phylogenetic backbones.}

% \noindent\includegraphics[width=0.85\linewidth]{FigureS2.png}

\subsection*{Figure S3: All-site mutational spectra of major VOCs.}
(A) Mutational spectra of the ARG.
For each major VOC, the relative proportion of each mutation type was calculated from
the mutations occurring above the nodes with a Scorpio label associated with the VOC
(e.g. “Alpha-like” encompasses “B.1.1.7-like” and “B.1.1.7-like+E484K”).
Note that ~98\% of all the mutations in the ARG occur above the nodes
which have a Scorpio label associated with Alpha, Delta, Omicron BA.1, BA.2, BA.4, or BA.5
(excluding the nodes with unassigned scorpio labels).
The Scorpio labels were assigned to the aligned sequences of the nodes in the ARG using pangolin.
(B) Mutational spectra reproduced from the mutation count data analyzed in Bloom et al. (2023).
The authors extracted mutation counts from a global UShER phylogeny (https://github.com/jbloomlab/SARS2-mut-spectrum/).

% \noindent\includegraphics[width=0.85\linewidth]{FigureS3.png}

\end{document}
