{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "177e62e6-7992-4946-913b-dc17ae730cae",
   "metadata": {},
   "source": [
    "# Time travellers in Viridian\n",
    "\n",
    "In this notebook we do some brief analysis on the time travelling samples (those with a collection date that is much earlier than their true collection date) in the Viridian dataset.\n",
    "\n",
    "This also provides a useful example of how to extract and manipulate metadata fields from the VCF Zarr encoding using Pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5fafd3e2-199a-4c61-b055-34b726aaebe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sc2ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dd27aba2-87e9-422b-8b30-07c8da3bda60",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "Dataset at ../data/viridian_mafft_2024-10-14_v1.vcz.zip with 4484157 samples, 29903 variants, and 30 metadata fields. See ds.metadata.field_descriptors() for a description of the fields."
      ],
      "text/plain": [
       "<sc2ts.dataset.Dataset at 0x7ffa2d02f070>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds = sc2ts.Dataset(\"../data/viridian_mafft_2024-10-14_v1.vcz.zip\")\n",
    "ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "84480b22-bf74-4cd2-892f-71ecfb70b73a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dtype</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>field</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Artic_primer_version</th>\n",
       "      <td>object</td>\n",
       "      <td>If known, the ARTIC primer scheme from ENA met...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Collection_date</th>\n",
       "      <td>object</td>\n",
       "      <td>collection_date from ENA metadata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <td>object</td>\n",
       "      <td>Country from ENA metadata. If ':' was in the e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date_tree</th>\n",
       "      <td>object</td>\n",
       "      <td>A consensus date, using up 3 sources of data f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date_tree_order</th>\n",
       "      <td>object</td>\n",
       "      <td>This helped define the order in which the samp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Experiment</th>\n",
       "      <td>object</td>\n",
       "      <td>experiment_accession from ENA metadata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>First_created</th>\n",
       "      <td>object</td>\n",
       "      <td>first_created from the ENA metadata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genbank_N</th>\n",
       "      <td>int16</td>\n",
       "      <td>Number of Ns in the GenBank consensus sequence...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genbank_accession</th>\n",
       "      <td>object</td>\n",
       "      <td>This is the GenBank accession of the assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genbank_other_runs</th>\n",
       "      <td>object</td>\n",
       "      <td>Any other run accessions associated with the G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genbank_pangolin</th>\n",
       "      <td>object</td>\n",
       "      <td>Pangolin call from the GenBank consensus seque...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genbank_scorpio</th>\n",
       "      <td>object</td>\n",
       "      <td>Scorpio call from the GenBank consensus sequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genbank_tree_name</th>\n",
       "      <td>object</td>\n",
       "      <td>Name of the sample in the GenBank tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>In_Viridian_tree</th>\n",
       "      <td>bool</td>\n",
       "      <td>'T' or 'F' to indicate if the sample is in the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>In_intersection</th>\n",
       "      <td>bool</td>\n",
       "      <td>'T' or 'F' to indicate if the sample is in the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>In_may_2024_preprint</th>\n",
       "      <td>bool</td>\n",
       "      <td>Whether or not this run was in the May 2024 pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Platform</th>\n",
       "      <td>object</td>\n",
       "      <td>instrument_platform from ENA metadata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <td>object</td>\n",
       "      <td>Region from ENA metadata, if present. If ':' w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run_count</th>\n",
       "      <td>int8</td>\n",
       "      <td>The number of runs from the sample accession</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <td>object</td>\n",
       "      <td>sample_accession from ENA metadata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study</th>\n",
       "      <td>object</td>\n",
       "      <td>study_accession from ENA metadata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_N</th>\n",
       "      <td>int16</td>\n",
       "      <td>Number of Ns in the viridian consensus sequenc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_amplicon_scheme</th>\n",
       "      <td>object</td>\n",
       "      <td>Amplicon scheme called by Viridian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_cons_het</th>\n",
       "      <td>int8</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_cons_len</th>\n",
       "      <td>int16</td>\n",
       "      <td>Length of the Viridian consensus sequence Viri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_pangolin</th>\n",
       "      <td>object</td>\n",
       "      <td>Pangolin call from the Viridian consensus sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_pangolin_1.29</th>\n",
       "      <td>object</td>\n",
       "      <td>Pangolin call from the Viridian consensus sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_result</th>\n",
       "      <td>object</td>\n",
       "      <td>This is 'PASS' if Viridian finished and made a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_scorpio</th>\n",
       "      <td>object</td>\n",
       "      <td>Scorpio call from the Viridian consensus seque...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_scorpio_1.29</th>\n",
       "      <td>object</td>\n",
       "      <td>Scorpio call from the Viridian consensus seque...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           dtype  \\\n",
       "field                              \n",
       "Artic_primer_version      object   \n",
       "Collection_date           object   \n",
       "Country                   object   \n",
       "Date_tree                 object   \n",
       "Date_tree_order           object   \n",
       "Experiment                object   \n",
       "First_created             object   \n",
       "Genbank_N                  int16   \n",
       "Genbank_accession         object   \n",
       "Genbank_other_runs        object   \n",
       "Genbank_pangolin          object   \n",
       "Genbank_scorpio           object   \n",
       "Genbank_tree_name         object   \n",
       "In_Viridian_tree            bool   \n",
       "In_intersection             bool   \n",
       "In_may_2024_preprint        bool   \n",
       "Platform                  object   \n",
       "Region                    object   \n",
       "Run_count                   int8   \n",
       "Sample                    object   \n",
       "Study                     object   \n",
       "Viridian_N                 int16   \n",
       "Viridian_amplicon_scheme  object   \n",
       "Viridian_cons_het           int8   \n",
       "Viridian_cons_len          int16   \n",
       "Viridian_pangolin         object   \n",
       "Viridian_pangolin_1.29    object   \n",
       "Viridian_result           object   \n",
       "Viridian_scorpio          object   \n",
       "Viridian_scorpio_1.29     object   \n",
       "\n",
       "                                                                description  \n",
       "field                                                                        \n",
       "Artic_primer_version      If known, the ARTIC primer scheme from ENA met...  \n",
       "Collection_date                           collection_date from ENA metadata  \n",
       "Country                   Country from ENA metadata. If ':' was in the e...  \n",
       "Date_tree                 A consensus date, using up 3 sources of data f...  \n",
       "Date_tree_order           This helped define the order in which the samp...  \n",
       "Experiment                           experiment_accession from ENA metadata  \n",
       "First_created                           first_created from the ENA metadata  \n",
       "Genbank_N                 Number of Ns in the GenBank consensus sequence...  \n",
       "Genbank_accession             This is the GenBank accession of the assembly  \n",
       "Genbank_other_runs        Any other run accessions associated with the G...  \n",
       "Genbank_pangolin          Pangolin call from the GenBank consensus seque...  \n",
       "Genbank_scorpio           Scorpio call from the GenBank consensus sequen...  \n",
       "Genbank_tree_name                    Name of the sample in the GenBank tree  \n",
       "In_Viridian_tree          'T' or 'F' to indicate if the sample is in the...  \n",
       "In_intersection           'T' or 'F' to indicate if the sample is in the...  \n",
       "In_may_2024_preprint      Whether or not this run was in the May 2024 pr...  \n",
       "Platform                              instrument_platform from ENA metadata  \n",
       "Region                    Region from ENA metadata, if present. If ':' w...  \n",
       "Run_count                      The number of runs from the sample accession  \n",
       "Sample                                   sample_accession from ENA metadata  \n",
       "Study                                     study_accession from ENA metadata  \n",
       "Viridian_N                Number of Ns in the viridian consensus sequenc...  \n",
       "Viridian_amplicon_scheme                 Amplicon scheme called by Viridian  \n",
       "Viridian_cons_het                                                            \n",
       "Viridian_cons_len         Length of the Viridian consensus sequence Viri...  \n",
       "Viridian_pangolin         Pangolin call from the Viridian consensus sequ...  \n",
       "Viridian_pangolin_1.29    Pangolin call from the Viridian consensus sequ...  \n",
       "Viridian_result           This is 'PASS' if Viridian finished and made a...  \n",
       "Viridian_scorpio          Scorpio call from the Viridian consensus seque...  \n",
       "Viridian_scorpio_1.29     Scorpio call from the Viridian consensus seque...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds.metadata.field_descriptors()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c02101bd-a8af-47cc-b777-b725038092ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date_tree</th>\n",
       "      <th>Viridian_scorpio_1.29</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR11772659</th>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11597132</th>\n",
       "      <td>2020-01-29</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12162233</th>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12162234</th>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11597217</th>\n",
       "      <td>2020-02-02</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177478</th>\n",
       "      <td>2023-04-20</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177479</th>\n",
       "      <td>2023-04-21</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177480</th>\n",
       "      <td>2023-04-24</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177481</th>\n",
       "      <td>2023-04-26</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177482</th>\n",
       "      <td>2023-04-27</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4484157 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Date_tree Viridian_scorpio_1.29\n",
       "sample_id                                    \n",
       "SRR11772659  2020-01-19                     .\n",
       "SRR11597132  2020-01-29                     .\n",
       "SRR12162233  2020-01-30                     .\n",
       "SRR12162234  2020-01-30                     .\n",
       "SRR11597217  2020-02-02                     .\n",
       "...                 ...                   ...\n",
       "ERR13177478  2023-04-20    Omicron (XBB-like)\n",
       "ERR13177479  2023-04-21    Omicron (XBB-like)\n",
       "ERR13177480  2023-04-24    Omicron (XBB-like)\n",
       "ERR13177481  2023-04-26    Omicron (XBB-like)\n",
       "ERR13177482  2023-04-27    Omicron (XBB-like)\n",
       "\n",
       "[4484157 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = ds.metadata.as_dataframe([\"Date_tree\", \"Viridian_scorpio_1.29\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b082ec4d-2913-4d6f-9f23-f4eb9f1f8a49",
   "metadata": {},
   "source": [
    "Filter out incomplete dates for simplicity. Full ISO dates are 10 chars long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d032bc3f-b127-40fe-a90c-2f1c7673f85f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date_tree</th>\n",
       "      <th>Viridian_scorpio_1.29</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR11772659</th>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11597132</th>\n",
       "      <td>2020-01-29</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12162233</th>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12162234</th>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11597217</th>\n",
       "      <td>2020-02-02</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177478</th>\n",
       "      <td>2023-04-20</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177479</th>\n",
       "      <td>2023-04-21</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177480</th>\n",
       "      <td>2023-04-24</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177481</th>\n",
       "      <td>2023-04-26</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13177482</th>\n",
       "      <td>2023-04-27</td>\n",
       "      <td>Omicron (XBB-like)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4458826 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Date_tree Viridian_scorpio_1.29\n",
       "sample_id                                    \n",
       "SRR11772659  2020-01-19                     .\n",
       "SRR11597132  2020-01-29                     .\n",
       "SRR12162233  2020-01-30                     .\n",
       "SRR12162234  2020-01-30                     .\n",
       "SRR11597217  2020-02-02                     .\n",
       "...                 ...                   ...\n",
       "ERR13177478  2023-04-20    Omicron (XBB-like)\n",
       "ERR13177479  2023-04-21    Omicron (XBB-like)\n",
       "ERR13177480  2023-04-24    Omicron (XBB-like)\n",
       "ERR13177481  2023-04-26    Omicron (XBB-like)\n",
       "ERR13177482  2023-04-27    Omicron (XBB-like)\n",
       "\n",
       "[4458826 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[df[\"Date_tree\"].str.len() == 10]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "28fcdb46-c673-4688-8cad-11df8ad663bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date_tree</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_scorpio_1.29</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>.</th>\n",
       "      <td>2020-01-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A.23.1-like</th>\n",
       "      <td>2020-04-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A.23.1-like+E484K</th>\n",
       "      <td>2020-12-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV.1-like</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alpha (B.1.1.7-like)</th>\n",
       "      <td>2020-01-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.1.318-like</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.1.7-like+E484K</th>\n",
       "      <td>2020-12-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.617.1-like</th>\n",
       "      <td>2020-03-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.617.3-like</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Beta (B.1.351-like)</th>\n",
       "      <td>2020-05-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (AY.4-like)</th>\n",
       "      <td>2020-10-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (AY.4.2-like)</th>\n",
       "      <td>2020-12-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (B.1.617.2-like)</th>\n",
       "      <td>2020-05-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (B.1.617.2-like) +K417N</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epsilon (B.1.427-like)</th>\n",
       "      <td>2020-09-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epsilon (B.1.429-like)</th>\n",
       "      <td>2020-09-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eta (B.1.525-like)</th>\n",
       "      <td>2020-12-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gamma (P.1-like)</th>\n",
       "      <td>2020-05-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Iota (B.1.526-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lambda (C.37-like)</th>\n",
       "      <td>2021-01-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mu (B.1.621-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.1-like)</th>\n",
       "      <td>2020-09-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.2-like)</th>\n",
       "      <td>2020-01-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.3-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.4-like)</th>\n",
       "      <td>2020-07-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.5-like)</th>\n",
       "      <td>2020-03-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (Unassigned)</th>\n",
       "      <td>2020-06-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB-like)</th>\n",
       "      <td>2020-10-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB.1-like)</th>\n",
       "      <td>2020-12-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB.1.16-like)</th>\n",
       "      <td>2020-11-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB.1.5-like)</th>\n",
       "      <td>2022-01-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XE-like)</th>\n",
       "      <td>2022-01-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (BA.1-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (BA.2-like)</th>\n",
       "      <td>2021-12-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (BA.3-like)</th>\n",
       "      <td>2022-01-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (BA.5-like)</th>\n",
       "      <td>2022-06-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (Unassigned)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (XBB-like)</th>\n",
       "      <td>2022-10-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (XBB.1-like)</th>\n",
       "      <td>2022-12-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (XBB.1.16-like)</th>\n",
       "      <td>2023-07-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (XBB.1.5-like)</th>\n",
       "      <td>2023-01-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Theta (P.3-like)</th>\n",
       "      <td>2021-02-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zeta (P.2-like)</th>\n",
       "      <td>2020-05-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Date_tree\n",
       "Viridian_scorpio_1.29                       \n",
       ".                                 2020-01-01\n",
       "A.23.1-like                       2020-04-12\n",
       "A.23.1-like+E484K                 2020-12-26\n",
       "AV.1-like                         2020-12-31\n",
       "Alpha (B.1.1.7-like)              2020-01-12\n",
       "B.1.1.318-like                    2020-12-31\n",
       "B.1.1.7-like+E484K                2020-12-08\n",
       "B.1.617.1-like                    2020-03-17\n",
       "B.1.617.3-like                    2020-12-31\n",
       "Beta (B.1.351-like)               2020-05-27\n",
       "Delta (AY.4-like)                 2020-10-27\n",
       "Delta (AY.4.2-like)               2020-12-21\n",
       "Delta (B.1.617.2-like)            2020-05-08\n",
       "Delta (B.1.617.2-like) +K417N     2020-12-31\n",
       "Epsilon (B.1.427-like)            2020-09-17\n",
       "Epsilon (B.1.429-like)            2020-09-25\n",
       "Eta (B.1.525-like)                2020-12-15\n",
       "Gamma (P.1-like)                  2020-05-06\n",
       "Iota (B.1.526-like)               2020-12-31\n",
       "Lambda (C.37-like)                2021-01-29\n",
       "Mu (B.1.621-like)                 2020-12-31\n",
       "Omicron (BA.1-like)               2020-09-22\n",
       "Omicron (BA.2-like)               2020-01-10\n",
       "Omicron (BA.3-like)               2020-12-31\n",
       "Omicron (BA.4-like)               2020-07-02\n",
       "Omicron (BA.5-like)               2020-03-31\n",
       "Omicron (Unassigned)              2020-06-30\n",
       "Omicron (XBB-like)                2020-10-23\n",
       "Omicron (XBB.1-like)              2020-12-22\n",
       "Omicron (XBB.1.16-like)           2020-11-20\n",
       "Omicron (XBB.1.5-like)            2022-01-17\n",
       "Omicron (XE-like)                 2022-01-19\n",
       "Probable Omicron (BA.1-like)      2020-12-31\n",
       "Probable Omicron (BA.2-like)      2021-12-30\n",
       "Probable Omicron (BA.3-like)      2022-01-28\n",
       "Probable Omicron (BA.5-like)      2022-06-11\n",
       "Probable Omicron (Unassigned)     2020-12-31\n",
       "Probable Omicron (XBB-like)       2022-10-04\n",
       "Probable Omicron (XBB.1-like)     2022-12-12\n",
       "Probable Omicron (XBB.1.16-like)  2023-07-03\n",
       "Probable Omicron (XBB.1.5-like)   2023-01-07\n",
       "Theta (P.3-like)                  2021-02-01\n",
       "Zeta (P.2-like)                   2020-05-04"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg = df.groupby(\"Viridian_scorpio_1.29\").min()\n",
    "dfg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e509fbcf-853f-4f6d-8e89-c45caf713b0e",
   "metadata": {},
   "source": [
    "As a very rough proxy for the existance of time travellers, we just look at the minimum date observed within each of the Scorpio designations. Nearly all of these are post 2020, but yet most have samples in 2020."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6746f6ce-5fd4-4e8d-bee2-2a10ecdb0ec7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfg_2020 = dfg[dfg[\"Date_tree\"] < \"2021\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "76c6a885-595d-4703-98ad-b929c2d3ac34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(32, 1)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg_2020.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ad4bb4b1-6444-4d07-ae53-0633efbdec08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43, 1)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c4267b32-b854-476f-ad49-147cb9d18b60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date_tree</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Viridian_scorpio_1.29</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>.</th>\n",
       "      <td>2020-01-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A.23.1-like</th>\n",
       "      <td>2020-04-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A.23.1-like+E484K</th>\n",
       "      <td>2020-12-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AV.1-like</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alpha (B.1.1.7-like)</th>\n",
       "      <td>2020-01-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.1.318-like</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.1.7-like+E484K</th>\n",
       "      <td>2020-12-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.617.1-like</th>\n",
       "      <td>2020-03-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.1.617.3-like</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Beta (B.1.351-like)</th>\n",
       "      <td>2020-05-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (AY.4-like)</th>\n",
       "      <td>2020-10-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (AY.4.2-like)</th>\n",
       "      <td>2020-12-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (B.1.617.2-like)</th>\n",
       "      <td>2020-05-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta (B.1.617.2-like) +K417N</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epsilon (B.1.427-like)</th>\n",
       "      <td>2020-09-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epsilon (B.1.429-like)</th>\n",
       "      <td>2020-09-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eta (B.1.525-like)</th>\n",
       "      <td>2020-12-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gamma (P.1-like)</th>\n",
       "      <td>2020-05-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Iota (B.1.526-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mu (B.1.621-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.1-like)</th>\n",
       "      <td>2020-09-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.2-like)</th>\n",
       "      <td>2020-01-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.3-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.4-like)</th>\n",
       "      <td>2020-07-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (BA.5-like)</th>\n",
       "      <td>2020-03-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (Unassigned)</th>\n",
       "      <td>2020-06-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB-like)</th>\n",
       "      <td>2020-10-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB.1-like)</th>\n",
       "      <td>2020-12-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Omicron (XBB.1.16-like)</th>\n",
       "      <td>2020-11-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (BA.1-like)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Probable Omicron (Unassigned)</th>\n",
       "      <td>2020-12-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zeta (P.2-like)</th>\n",
       "      <td>2020-05-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Date_tree\n",
       "Viridian_scorpio_1.29                    \n",
       ".                              2020-01-01\n",
       "A.23.1-like                    2020-04-12\n",
       "A.23.1-like+E484K              2020-12-26\n",
       "AV.1-like                      2020-12-31\n",
       "Alpha (B.1.1.7-like)           2020-01-12\n",
       "B.1.1.318-like                 2020-12-31\n",
       "B.1.1.7-like+E484K             2020-12-08\n",
       "B.1.617.1-like                 2020-03-17\n",
       "B.1.617.3-like                 2020-12-31\n",
       "Beta (B.1.351-like)            2020-05-27\n",
       "Delta (AY.4-like)              2020-10-27\n",
       "Delta (AY.4.2-like)            2020-12-21\n",
       "Delta (B.1.617.2-like)         2020-05-08\n",
       "Delta (B.1.617.2-like) +K417N  2020-12-31\n",
       "Epsilon (B.1.427-like)         2020-09-17\n",
       "Epsilon (B.1.429-like)         2020-09-25\n",
       "Eta (B.1.525-like)             2020-12-15\n",
       "Gamma (P.1-like)               2020-05-06\n",
       "Iota (B.1.526-like)            2020-12-31\n",
       "Mu (B.1.621-like)              2020-12-31\n",
       "Omicron (BA.1-like)            2020-09-22\n",
       "Omicron (BA.2-like)            2020-01-10\n",
       "Omicron (BA.3-like)            2020-12-31\n",
       "Omicron (BA.4-like)            2020-07-02\n",
       "Omicron (BA.5-like)            2020-03-31\n",
       "Omicron (Unassigned)           2020-06-30\n",
       "Omicron (XBB-like)             2020-10-23\n",
       "Omicron (XBB.1-like)           2020-12-22\n",
       "Omicron (XBB.1.16-like)        2020-11-20\n",
       "Probable Omicron (BA.1-like)   2020-12-31\n",
       "Probable Omicron (Unassigned)  2020-12-31\n",
       "Zeta (P.2-like)                2020-05-04"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg_2020"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
